utils/scarb-oracle-hint-service/src/connections/shell.rs

```diff
@@ -0,0 +1,157 @@
+use crate::Protocol;
+use crate::connection::Connection;
+use anyhow::{Context, Result, anyhow, ensure};
+use deno_task_shell::parser::{SequentialList, parse};
+use deno_task_shell::{ShellPipeReader, ShellState, pipe};
+use starknet_core::codec::{Decode, Encode};
+use starknet_core::types::{ByteArray, Felt};
+use std::env;
+use std::io::{BufRead, BufReader, Read};
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use tokio::task::spawn_blocking;
+use tracing::{Span, debug, debug_span, warn};
+
+/// The `shell` protocol executes a provided shell invocation line and returns its stdout as
+/// a `ByteArray`. Stderr is redirected to the logs.
+pub struct Shell;
+
+impl Protocol for Shell {
+    const SCHEME: &'static str = "shell";
+
+    #[tracing::instrument]
+    fn connect(_command: &str) -> Result<Box<dyn Connection + 'static>> {
+        Ok(Box::new(ShellConnection::new()?))
+    }
+}
+
+struct ShellConnection {
+    rt: tokio::runtime::Runtime,
+}
+
+impl ShellConnection {
+    fn new() -> Result<Self> {
+        Ok(Self {
+            rt: tokio::runtime::Builder::new_current_thread().build()?,
+        })
+    }
+}
+
+impl Connection for ShellConnection {
+    fn call(&mut self, selector: &str, calldata: &[Felt]) -> Result<Vec<Felt>> {
+        // This oracle has several modes which are determined by suffixes of the selector.
+        // All selectors accept the same input format.
+        let mode = Mode::get(selector)?;
+
+        let span = Arc::new(debug_span!("exec", id = next_id()));
+        let _enter = span.enter();
+
+        // Decode arguments.
+        let mut calldata = calldata.iter();
+        let command: String = ByteArray::decode_iter(&mut calldata)?.try_into()?;
+        debug!("{command}");
+
+        // Run command.
+        let command = parse(&command)?;
+        let shell = build_shell()?;
+        let (exit_code, stdout) = self.rt.block_on(execute(span.clone(), command, shell));
+
+        // Finalise, according to mode.
+        mode.finalise(exit_code, stdout.into())
+    }
+}
+
+#[derive(Eq, PartialEq)]
+enum Mode {
+    /// `eo`: return (exit code, stdout) tuple.
+    ReturnExitReturnOutput,
+    /// `co`: require successful exit code, return stdout.
```

Comment:
What does the `co` stand for? 🤔 
This seem kind of confusing for me, but I guess the flexibility is worth it.